### ISSUE TITLE
Expose underlying Circuit Breaker implementation instance to allow for additional configuration

### ISSUE THREAD

**Issue by madorb** [spring-cloud/spring-cloud-circuitbreaker#59]
**Is your feature request related to a problem? Please describe.**
While it certainly is a perfectly defensible position that SCCB (as per its name) should only support the "CircuitBreaker" pattern - it may be more robust to allow for configuration of additional library-specific resilience patterns. For instance Resilience4j enables 7 different [resilience patterns](https://github.com/resilience4j/resilience4j#resilience-patterns) of which only CircuitBreaker and TimeLimiter are [supported](https://github.com/spring-cloud/spring-cloud-circuitbreaker/blob/master/spring-cloud-circuitbreaker-resilience4j/src/main/java/org/springframework/cloud/circuitbreaker/resilience4j/Resilience4JConfigBuilder.java#L32)

**Describe the solution you'd like**
Would love to have a mechanism to more fully configure the underlying circuit breaker's implementation, for instance to also enable Retry.

**Comment by ryanjbaxter** [comment]
I think retry can be implemented pretty easily outside of Spring Cloud CircuitBreaker, do you have a good reason as to why it should be part of the API?

**Comment by spring-issuemaster** [comment]
If you would like us to look at this issue, please provide the requested information. If the information is not provided within the next 7 days this issue will be closed.

**Comment by madorb** [comment]
Sorry for the delayed response, yes Retry can be implemented relatively easily outside of SCCB, but lets say I want to use Resilience4j for retry as well. That means I have two separate beans using two separate spring abstractions (retry and circuitbreaker) around a single http call. 

What about if i DON'T want to use retry but want to use, say, RateLimiter, or Bulkhead [resilience patterns](https://github.com/resilience4j/resilience4j#resilience-patterns)? Spring has no abstraction for these, but `Resilience4j` supports them. This means that if I want to use CircuitBreaker/TimeLimiter AND _any_ other resilience patterns, there's no value to me in using SCCB at all, as I still have to configure my own R4j instance and use it manually to configure the other resilience patterns. 

Alternately, SCCB could provide a mechanism to natively configure the underlying CircuitBreaker provider - I think keeping the abstraction at the level of CircuitBreaker makes sense, and it can be unaware of the details of the other resiliency patterns I might want to use, but it might be useful to at least allow me to configure them.

Just a thought, obviously there are other potential solutions for this (i.e. avoiding the Spring Cloud abstraction completely, and just use the native [boot support](https://resilience4j.readme.io/docs/getting-started-3) of resilience4j, though i think this may limit the usefulness of having the abstraction in the first place)

Thanks for considering!!

**Comment by breun** [comment]
Also the import of `resilience4j-bom` is currently done in `spring-cloud-circuitbreaker-dependencies`, but if you want to use any of the Resilience4J modules like bulkhead or ratelimiter, but don't really have a need for Spring Cloud Circuit Breaker, then you still need to depend on `spring-cloud-circuitbreaker-dependencies`, or you need to import `resilience4j-bom` yourself and manage its version as well.

I've created https://github.com/spring-cloud/spring-cloud-circuitbreaker/issues/91 to propose moving the import of `resilience4j-bom` up to `spring-cloud-dependencies`, so those other managed Resilience4j modules can also be imported by users that don't use Spring Cloud Circuit Breaker.

**Comment by Richardmbs12** [comment]
I also think this would make a lot of sense as a feature.

The reason I am using SCCB is because of the abstraction. If I anyway need to make use of resilience4j retry mechanism for example, which is not part of the SCCB abstraction, then I could have just gone with plain resilience4j and not care about the abstraction SCBB provides, as I am already coupling myself to resilience4j if I want the retry feature.

Retrying also makes sense in the circuit breaker context, especially when we want to be able to tune our circuit breakers, not to increase the fail rate before retrying first

**Comment by ryanjbaxter** [comment]
The one concern I have with supporting other resiliency patterns is supporting them across all implementations.  Yes R4J supports many different patterns but what about the other libraries?

What about Sentinal and Spring Retry?

We need to support those implementations as well.

So lets open specific issues for each pattern we want to support and then we can come up with ways of implementing it across all implementations.

**Comment by madorb** [comment]
@ryanjbaxter i totally agree about that approach for supporting differing reliability patterns... that doesn't solve my initial request though - to make the underlying CB provider available to be further configured...

**Comment by spring-cloud-issues** [comment]
If you would like us to look at this issue, please provide the requested information. If the information is not provided within the next 7 days this issue will be closed.

**Comment by spring-cloud-issues** [comment]
Closing due to lack of requested feedback. If you would like us to look at this issue, please provide the requested information and we will re-open the issue.
